lib/api/api-fetch/body.js

```diff
@@ -1,66 +1,178 @@
     }
 
-    let stream
-    if (util.isStream(body)) {
-      stream = Readable.toWeb(body)
-    } else {
-      if (body.locked) {
-        throw new TypeError('locked')
+    if (body.locked) {
+      throw new TypeError('locked')
+    }
+
+    stream = body
+  } else {
+    source = String(body)
+    length = Buffer.byteLength(source)
+    type = 'text/plain;charset=UTF-8'
+  }
+
+  if (!stream) {
+    assert(source != null)
+    if (!ReadableStream) {
+      ReadableStream = require('stream/web').ReadableStream
+    }
+    stream = new ReadableStream({
+      async start (controller) {
+        controller.enqueue(source)
+        controller.close()
+      },
+      async pull () {
+      },
+      async cancel (reason) {
       }
+    })
+  }
+
+  return [{ stream, source, length }, type]
+}
 
-      if (!TransformStream) {
-        TransformStream = require('stream/web').TransformStream
+function cloneBody (src) {
+  if (!src) {
+    return
+  }
+
+  if (util.isDisturbed(src)) {
+    throw new TypeError('disturbed')
+  }
+
+  if (src.locked) {
+    throw new TypeError('locked')
+  }
+
+  // https://fetch.spec.whatwg.org/#concept-body-clone
+  const [out1, out2] = src.stream.tee()
+  src.stream = out1
+  return {
+    stream: out2,
+    length: src.length,
+    source: src.source
+  }
+}
+
+function safelyExtractBody (body) {
+  if (util.isDisturbed(body)) {
+    throw new TypeError('disturbed')
+  }
+
+  if (body && body.locked) {
+    throw new TypeError('locked')
+  }
+
+  return extractBody(body)
+}
+
+const methods = {
+  async blob () {
+    const chunks = []
+
+    if (this[kState].body) {
+      if (this.bodyUsed || this[kState].body.stream.locked) {
+        throw new TypeError('unusable')
       }
 
-      // https://streams.spec.whatwg.org/#readablestream-create-a-proxy
-      const identityTransform = new TransformStream()
-      body.pipeThrough(identityTransform)
-      stream = identityTransform
+      this[kState].bodyUsed = true
+      for await (const chunk of this[kState].body.stream) {
+        chunks.push(chunk)
+      }
     }
 
-    return [{
-      stream
-    }, null]
-  } else {
-    throw Error('Cannot extract Body from input: ', body)
+    return new Blob(chunks, { type: this.headers.get('Content-Type') || '' })
+  },
+
+  async arrayBuffer () {
+    const blob = await this.blob()
+    return await blob.arrayBuffer()
+  },
+
+  async text () {
+    const blob = await this.blob()
+    return await blob.text()
+  },
+
+  async json () {
+    return JSON.parse(await this.text())
+  },
+
+  async formData () {
+    // TODO: Implement.
+    throw new NotSupportedError('formData')
+  }
+}
+
+const properties = {
+  body: {
+    enumerable: true,
+    get () {
+      return this[kState].body ? this[kState].body.stream : null
+    }
+  },
+  bodyUsed: {
+    enumerable: true,
+    get () {
+      return this[kState].body && (
+        util.isDisturbed(this[kState].body.stream) ||
+        !!this[kState].bodyUsed
```

Comment:
why is part necessary?

Reply:
IsDistrubed doesn’t work yet. Haven’t been released on 16.x yet